### DISCUSSION TITLE
Custom Data Parser for spaCy 3.0

### DISCUSSION THREAD

**Discussion by davesuketu215** [explosion/spaCy#7001] [Help: Coding & Implementations]
I am implementing a custom data parser/streamer for my use-case.
I am referring to the same example given here - https://spacy.io/usage/training#custom-code-readers-batchers

While executing from CLI, I observe that the generate_stream( ) is called for a certain number of times, during the pipeline initialization as well as training. I am not able to understand why is it being called multiple times in an epoch.

Is there any reference document which gives a detailed outlook on the call stack that runs beneath once the training is fired from CLI? This will help us to understand the things better and design accordingly.

Also, I want to extract the training and validation set by splitting from the same set of examples doc objects.

**Comment by benbatorsky-ciox** [comment]
I think I'm having a similar issue.  I tracked mine down during the train iteration to here: https://github.com/explosion/spaCy/blob/4188beda871b1e40eb8d02b8a787b6878c89717e/spacy/training/loop.py#L294

It seems like it takes the iterator and extracts the whole thing all at once.  This seems like it would happen even if you wrote a custom batcher, as it happens before the batching.  But if I just run the batcher on my iterator, it seems like it'd work fine, without having to iterate through the WHOLE dataset at once.  Anyone have thoughts here?
